### ISSUE TITLE
would mimekit insert newline character after certain length? 

### ISSUE THREAD

**Issue by blacklion9279** [jstedfast/MimeKit#107]

**Comment by blacklion9279** [comment]
so we are noticing that a if a text part inside mimebody has really long line, in this case about 3500 characters long, approximately at 989 characters at inserts a newline

**Comment by jstedfast** [comment]
MailKit's SmtpClient and the MultipartSigned logic will change the Content-Transfer-Encoding of MimePart's to quoted-printable if there are long lines.

They both call MimePart's GetBestEncoding() method to figure that out.

**Comment by jstedfast** [comment]
The BodyBuilder API might also do it...

**Comment by blacklion9279** [comment]
thanks for that answer @jstedfast  so the wired thing I am seeing is it seems like exchange server when sending email to other server adds some new line characters. When a signed message is sent, it has that long line. when it's received in sent items for that user it doesn't contain line breaks. However, when the same email is received by some other recipient, on some other server, it has those line breaks. 

Mimekit correctly identifies signing verification failure when it sees the line breaks. Is there anything in MimeKit to account for such freak cases? 

see example here: 
![image](https://cloud.githubusercontent.com/assets/1731171/6641469/4ebbc2dc-c974-11e4-9da6-bf218ed0ba31.png)

**Comment by jstedfast** [comment]
Just to help clarify things for me...

Are you:

1. creating/signing messages with MimeKit and sending them through Exchange SMTP? And when that message is forwarded on to server X (whatever that may be), the recipient receives the message with line breaks even though it did not start off with them?
2. or are you sending/signing messages with some other client through Exchange and then receiving them with MimeKit at the final destination (where it may go through another server which wraps long lines)?

I'm not clear on what is going on.

**Comment by blacklion9279** [comment]
It's the first one. Creating / signing message with app that uses mimekit and sends using EAS protocol. When it starts it doesn't have line breaks. When the recipient receives on server x it has line breaks. 

on receiving end windows based clients are showing signature failure. Outlook and our client. I am going to test it on other clients like iOS to see the behavior.

**Comment by jstedfast** [comment]
I've just modified MimeKit's multipart/signed logic to use a max line length of 78 instead of 998 to maximize compatibility. Hopefully this resolves your issue.

**Comment by blacklion9279** [comment]
so using quoted printable instead of default seems to be fixing this issue. 

if (mime.IsBodyHtml)
            {
                mimeMessage.Body = new TextPart("html") { Text = mime.Body, ContentTransferEncoding = ContentEncoding.QuotedPrintable};
            }

it was 

if (mime.IsBodyHtml)
            {
                mimeMessage.Body = new TextPart("html") { Text = mime.Body};
            }

**Comment by jstedfast** [comment]
Correct. This is essentially what the MultipartSigned logic has been fixed to do (detect if there are any long lines and if so, force the encoding to QuotedPrintable).

Glad to hear that fixes things for you :-)

I plan to make a new NuGet release this weekend, probably on Sunday.

**Comment by blacklion9279** [comment]
sounds good. do you see any downside to forcing ContentEncoding.QuotedPrintable everytime? how did you come up with that number 78 or 998?

Thanks for helping. You are rockstar!!

**Comment by jstedfast** [comment]
Well, QuotedPrintable adds encoding/decoding overhead and also increases the size of the message and so preferably you wouldn't use it unless it's necessary. When cryptographically signing messages it's probably not that big of a deal to always do it, but I guess I just like perfection :-)

I originally used 998 because that is the maximum line length allowed by SMTP (total line length is actually 1000 but if you subtract the CRLF, that leaves you with 998). In [rfc0821](https://github.com/jstedfast/MailKit/blob/master/rfc/rfc0821.txt#L2550), you'll find:

```
            text line

               The maximum total length of a text line including the
               <CRLF> is 1000 characters (but not counting the leading
               dot duplicated for transparency).
```

As far as 78 goes, MIME was historically designed around the idea that people used 80-column terminals to read email and so it was strongly recommended that MIME implementations wrapped lines at around 78 characters or less. In [rfc2822](https://github.com/jstedfast/MimeKit/blob/master/rfc/rfc2822.txt#L308), you'll find:

```
2.1.1. Line Length Limits

   There are two limits that this standard places on the number of
   characters in a line. Each line of characters MUST be no more than
   998 characters, and SHOULD be no more than 78 characters, excluding
   the CRLF.

   The 998 character limit is due to limitations in many implementations
   which send, receive, or store Internet Message Format messages that
   simply cannot handle more than 998 characters on a line. Receiving
   implementations would do well to handle an arbitrarily large number
   of characters in a line for robustness sake. However, there are so
   many implementations which (in compliance with the transport
   requirements of [RFC2821]) do not accept messages containing more
   than 1000 character including the CR and LF per line, it is important
   for implementations not to create such messages.

   The more conservative 78 character recommendation is to accommodate
   the many implementations of user interfaces that display these
   messages which may truncate, or disastrously wrap, the display of
   more than 78 characters per line, in spite of the fact that such
   implementations are non-conformant to the intent of this
   specification (and that of [RFC2821] if they actually cause
   information to be lost). Again, even though this limitation is put on
   messages, it is encumbant upon implementations which display messages
   to handle an arbitrarily large number of characters in a line
   (certainly at least up to the 998 character limit) for the sake of
   robustness.
```

Hope that helps.

**Comment by blacklion9279** [comment]
great explanation! thanks helps a lot! 

so one question out of curiosity - if the parser encountered a line greater than 998 characters, why would it not specify Content-Transfer-Encoding: quoted-printable ?

**Comment by jstedfast** [comment]
The parser doesn't modify anything, so I suspect you meant the MultipartSigned signing logic?

It should have done that, if it didn't, I don't know why it wouldn't have... it was designed to use QuotedPrintable for any content that had lines longer than 998 characters. Now it's been changed to do that for anything with lines longer than 78 characters.

**Comment by blacklion9279** [comment]
thanks that makes sense. appreciate all the help!! great work!!

**Comment by jstedfast** [comment]
Thanks and no problem :-)

**Comment by jstedfast** [comment]
I just released MimeKit 1.0.10 to NuGet.
